src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -463,6 +470,315 @@ int RGWBucket::check_object_index(const DoutPrefixProvider *dpp,
+  }
+  if (!op_state.hide_progress) {
+    ldpp_dout(dpp, 1) << "NOTICE: finished all shards (" << *count_out <<
+      " entries " << verb << ")" << dendl;
+  }
+  if (op_state.dump_keys) {
+    formatter->close_section();
+    flusher.flush();
+  }
+  return 0;
+}
+
+static int is_versioned_instance_listable(const DoutPrefixProvider *dpp,
+                                          RGWRados::BucketShard& bs,
+                                          const cls_rgw_obj_key& key,
+                                          bool& listable,
+                                          optional_yield y) {
+  const std::string empty_prefix;
+  rgw_cls_list_ret result;
+  librados::ObjectReadOperation op;
+  cls_rgw_bucket_list_op(op, key, empty_prefix, empty_prefix, 1,
+                         true, &result);
+  bufferlist ibl;
+  int r = bs.bucket_obj.operate(dpp, &op, &ibl, y);
+  if (r < 0) {
+    return r;
+  }
+  listable = result.dir.m.size() == 1 && key == result.dir.m.begin()->second.key;
+  return 0;
+}
+
+int RGWBucket::check_index_unlinked(const DoutPrefixProvider *dpp, RGWBucketAdminOpState& op_state,
+                                    RGWFormatterFlusher& flusher, const int shard, 
+                                    uint64_t* const count_out, optional_yield y) {
+  string marker = BI_INSTANCE_ENTRY_NS_START;
+  bool is_truncated = true;
+  list<rgw_cls_bi_entry> entries;
+
+  rgw::sal::RadosStore* rados_store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+  if (!rados_store) {
+    return 0;
+  }
+  RGWRados* store = rados_store->getRados();
+  RGWObjectCtx obj_ctx(driver);
+
+  RGWRados::BucketShard bs(store);
+  int ret = bs.init(dpp, bucket->get_info(), get_bucket_info().layout.current_index, shard, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, -1) << "ERROR bs.init(bucket=" << bucket << "): " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  ceph::real_clock::time_point now = ceph::real_clock::now();
+  ceph::real_clock::time_point not_after = now - std::chrono::hours(op_state.min_age);
+  
+  *count_out = 0;
+  do {
+    entries.clear();
+    ret = store->bi_list(bs, "", marker, -1, &entries, &is_truncated, y);
+    if (ret < 0) {
+      ldpp_dout(dpp, -1) << "ERROR bi_list(): " << cpp_strerror(-ret) << dendl;
+      break;
+    }
+    list<rgw_cls_bi_entry>::iterator iter;
+    for (iter = entries.begin(); iter != entries.end(); ++iter) {
+      rgw_cls_bi_entry& entry = *iter;
+      marker = entry.idx;
+      if (entry.type != BIIndexType::Instance) {
+        is_truncated = false;
+        break;
+      }
+      rgw_bucket_dir_entry dir_entry;
+      auto iiter = entry.data.cbegin();
+      try {
+        decode(dir_entry, iiter);
+      } catch (buffer::error& err) {
+        ldpp_dout(dpp, -1) << "ERROR failed to decode instance entry for key: " <<
+          entry.idx << dendl;
+        continue;
+      }
+      if (dir_entry.versioned_epoch == 0 &&
+          dir_entry.meta.mtime < not_after) {
```

Comment:
technically, wouldn't "not after" correspond to `<=` here?

and similar feedback about using `continue` when filtering to reduce nesting below